test/integration_test_wrapper.py

```diff
@@ -0,0 +1,274 @@
+    per_test_output_dir = args.mkosi_output_path.parent / args.test_name
+    per_test_output_dir.parent.mkdir(parents=True, exist_ok=True)
+    log.debug(f"Copying {args.mkosi_output_path.absolute()} output dir "
+              f"to per-test path {per_test_output_dir.absolute()}")
+
+    hook = None
+    if args.hook_module is not None:
+        spec = spec_from_file_location('hook', args.hook_module)
+        hook = module_from_spec(spec)
+        sys_modules['hook'] = hook
+        spec.loader.exec_module(hook)
+
+    with ExitStack() as stack:
+        stack.enter_context(lock_file(per_test_output_dir))
+
+        with TemporaryDirectory(dir=per_test_output_dir.parent.absolute(),
+                                prefix=f".{per_test_output_dir.name}",
+                                suffix=".tmp") as td:
+            log.debug(f"Copying to temporary path {td}")
+            if geteuid() == 0:
+                cp_args = [
+                    'cp',
+                    '--archive',
+                    '--reflink=auto',
+                    '--no-target-directory',
+                    args.mkosi_output_path,
+                    td,
+                ]
+                if log.isEnabledFor(LOG_DEBUG):
+                    log.debug(
+                        "Using rootly copy of output directory with command "
+                        + ' '.join(quote(str(a)) for a in cp_args))
+                run(cp_args, check=True)
+            else:
+                log.debug("Using namespace aware copy of build directory")
+                for e in listdir(args.mkosi_output_path):
+                    src = args.mkosi_output_path / e
+                    dst = Path(td) / e
+                    cp_args = []
+
+                    if src.is_dir():
+                        cp_args += ['unshare', '--map-auto', '--map-root-user']
+
+                    args += ['cp', '--archive', '--reflink=auto', src, dst]
+                    run(cp_args, check=True)
+
+            log.debug(f"Cleaning old dir {per_test_output_dir.absolute()}")
+            with suppress(FileNotFoundError):
+                rmtree(per_test_output_dir)
+
+            log.debug(f"Renaming tempdir {td} onto "
+                      f"{per_test_output_dir.absolute()}")
+            rename(td, per_test_output_dir)
+
+        def run_mkosi(mkosi_args, run=run, **kwargs):
+            clargs = [
+                'mkosi',
+                '--directory=..',
+                '--machine', args.test_name.replace('/', '_')[:64],
+                '--image', args.mkosi_image_name,
+                '--format=disk',
+                '--output-dir', per_test_output_dir.absolute(),
+            ] + mkosi_args
+            if log.isEnabledFor(LOG_DEBUG):
+                log.debug(f"Running {' '.join(quote(str(a)) for a in clargs)}")
+
+            return run(clargs, **kwargs)
+
+        machine_id = sha256(args.test_name.encode()).hexdigest()[:32]
+        log.debug(f"Derived machine-id {machine_id} for VM")
```

Comment:
Do we need to derive the machine ID from the test? The hostname will already be set to the test name, I don't think we need a derived machine ID.

Reply:
When I was forwarding logs out with the journald remote services I needed a consistent machine-id to find the right one for the right system. Explicit journal forwarding will remove the need for it.